### DISCUSSION TITLE
Kh8 hybrid modbus help using esp32

### DISCUSSION THREAD

**Discussion by iqbaal77** [nathanmarlor/foxess_modbus#664] [Support]
Doc: https://github.com/nathanmarlor/foxess_modbus 
https://github.com/nathanmarlor/foxess_modbus/wiki/Modbus-Wiring-Guide
https://github.com/nathanmarlor/foxess_modbus/files/11552447/KH.Series.Modbus.pdf
https://github.com/nathanmarlor/foxess_modbus/files/11511264/Energy.Storage.Inverter.Modbus.TCP.Communication.protocols.for.FOXESS-1.pdf

I am beginner trying to get data from foxess kh8 inverter but getting very long data in decimal or HEX using below code, can anyone help me to parse it or get correct readable form or this response is correct or not or how I can get correct response?

rs485 to max485 to esp-wroom-32 ch340 kit
inverter modbus address: 247

I connected:
max485 -> esp32
RO -> RX
(DE + RE common)  -> 5 pin
DI -> TX
GND -> GND

My code: 
```c
#include <Arduino.h>

#define RX 16      // UART2 RX
#define TX 17      // UART2 TX
#define RTS_pin 5  // RS485 Direction control
#define RS485Transmit HIGH
#define RS485Receive LOW

void setup() {
  pinMode(RTS_pin, OUTPUT);
  digitalWrite(RTS_pin, RS485Receive);

  // Start the Modbus serial Port
  Serial2.begin(9600, SERIAL_8N1, RX, TX);
  Serial.begin(9600);  // Ensure Serial Monitor is set to 9600
  delay(1000);

  Serial.println("Test RS485 Connection");

  while (Serial2.available()) {
    Serial.print(Serial2.read(), HEX);
    Serial.print(" ");
  }

  delay(1000);
}

uint16_t Calc_CRC(uint8_t* message, uint8_t len) {
  uint16_t crc = 0xFFFF;
  for (int pos = 0; pos < len; pos++) {
    crc ^= (uint16_t)message[pos];
    for (int i = 8; i != 0; i--) {
      if ((crc & 0x0001) != 0) {
        crc >>= 1;
        crc ^= 0xA001;
      } else {
        crc >>= 1;
      }
    }
  }
  return crc;
}

String Hex2String(uint8_t num) {
  char hexCar[2];
  sprintf(hexCar, "0x%02X", num);
  return hexCar;
}

void loop() {
  byte request[] = {0xF7 0x03 0x00 0x00 0x00 0x07 0x10 0x9E };

  uint16_t crc = Calc_CRC(request, sizeof(request) - 2);
  request[sizeof(request) - 1] = highByte(crc);
  request[sizeof(request) - 2] = lowByte(crc);

  Serial.println("Schreibe Daten ....");
  for (uint8_t i = 0; i < sizeof(request); i++) {
    Serial.print(Hex2String(request[i]));
    Serial.print(' ');
  }
  Serial.println("");

  digitalWrite(RTS_pin, RS485Transmit);
  Serial2.write(request, sizeof(request));
  Serial2.flush();
  digitalWrite(RTS_pin, RS485Receive);

  delay(100);
  Serial.println("Lese Daten ....");

  while (Serial2.available()) {
    Serial.print(Serial2.read(), HEX);
    Serial.print(" ");
  }
  Serial.println("");
  delay(2000);
}
```

Serial2.read response:
Schreibe Daten ....
0xF7 0x03 0x00 0x00 0x00 0x07 0x10 0x9E 
Lese Daten ....
86 EE 0 0 0 7 1F 1E 
or

Schreibe Daten ....
0xF7 0x03 0x00 0x00 0x00 0x07 0x10 0x9E 
Lese Daten ....
0 0 0 0 6 0 1 1 C0 E0 E0 60 E1 60 D0 10 70 50 40 40 E0 F1 60 F0 70 60 10 0 70 20 30 28 18 18 18 C C 10 1A 1C E 4 5 2 4 80 E0 C0 60 C0 F0 18 38 18 C E 2 6 4 3 1 E0 C0 E0 54 B0 30 20 70 58 18 1C 18 1C A 2 E E 3 2 C4 B8 0 C0 C0 60 B0 30 78 B0 28 1A C F 60 E0 28 30 18 58 B 1C 8 C 2 B 6 6 E E 7 C0 E0 60 30 38 10 30 38 18 19 8 18 18 A C0 60 30 30 78 98 18 8 4 6 2 E 6 6 7 6 6 42 C0 30 70 20 18 18 C 6 4 3 C0 E0 60 40 20 30 16 0 4C 6 6 7 7 80 C0 60 60 60 30 32 70 90 10 70 70 38 30 18 14 4C 2 2 7 1 1 C0 C0 E0 D0 70 38 0 C 4 7 87 3 1 44 E0 A8 60 70 31 78 30 8 8 A 7 3 1 C0 A0 E0 74 B8 18 4 48 98 C C E 7 3 80 E0 50 D0 40 60 30 70 38 20 30 28 38 38 30 18 10 18 14 1C 6 3 C0 C0 F0 61 10 30 38 30 78 E4 30 B0 F4 E0 E0 E0 F0 F7 E5 F0 A0 F8 E0 E1 E0 F0 E8 F0 F0 F8 78 78 F8 E8 60 F0 B0 E1 E0 F4 F0 F4 F8 34 70 FA 78 7C FE 58 7A FC 7E 78 70 FA F0 7C FC F4 78 F2 F9 E8 FB 9C 7F BE 1C 7E 3F FF 3F 9E 3E C7 BD 7F FF FF FF 9F FB 7E FF BF FF 7F FF DF FF FF FF FF FF BF FF FF FF 7F FF FF FF FF FF FF FF FF BF FF 0 3 0 0 0 0 0 0 6 2 0 0 2 0 0 0 0

**Comment by canton7** [comment]
There are a few issues with your code, but nothing which should stop it working for small frames, as far as I can see. None of the data you got back looks like a valid modbus frame though. 

Try just enabling the pull-up resistor on your rx line? It's going high-z when you transmit, and a floating voltage might be confusing the UART peripheral a bit, but nothing that should lead to a large amount of garbage.

Are you sure you connected to the correct pins on the inverter?

**Reply by FozzieUK** [reply]
> 
> Are you sure you connected to the correct pins on the inverter?

I was thinking the same, it looks more like a meter request

**Reply by canton7** [reply]
I wondered the same (and it would nicely explain why the amount of data is variable), but... I would have thought that the request to the meter would be fairly small (it's only asking for some values to be read, I think?), and while it might be the response from a meter, I would have thought OP would have noticed if they'd attached their RS485 wires to some pins which already had meter wires in...

**Reply by iqbaal77** [reply]
Last i tried this

This as well because I'm not sure which one is correct!
0xF7 0x03 0x00 0x00 0x00 0x07 0x10 0x9E

0xF7 0x04 0x00 0x00 0x00 0x07 0x10 0x9E

0xF7, 0x04, 0x2A, 0xF8, 0x00, 0x07, 0x00, 0x00

And also 
Schreibe Daten ....
0xF7 0x04 0x00 0x00 0x00 0x07 0xA5 0x5E 
Lese Daten ....
84 80 0 0 F A5 FF 
Schreibe Daten ....
0xF7 0x04 0x00 0x00 0x00 0x07 0xA5 0x5E 
Lese Daten ....
6 80 0 0 F9 DF 
Schreibe Daten ....
0xF7 0x04 0x00 0x00 0x00 0x07 0xA5 0x5E 
Lese Daten ....
86 80 80 0 E8 FF 
Schreibe Daten ....
0xF7 0x04 0x00 0x00 0x00 0x07 0xA5 0x5E 
Lese Daten ....
86 0 80 80 D8 BE

**Reply by canton7** [reply]
I'm reasonably sure that, whatever you're reading, it isn't a response from the inverter.

**Reply by iqbaal77** [reply]
I 100% connected with rs485A and rs485B of kh8 hybrid inverter

**Reply by canton7** [reply]
I'd start by double-checking everything electrical. Correct pins? Correct supply to the max485? Correct connection to the inverter? Terminating resistor? Correct voltage on the RS485 bus when transmitting?

**Reply by canton7** [reply]
Then I'd get a logic analyzer out and look at those serial RX and TX lines. When does the RX data actually come in? Is it continuous or in response to your request, with a consistent timing? Does it look generally well-formed, or are there invalid frames etc in there?

**Reply by iqbaal77** [reply]
Esp32 is getting voltage from my laptop usb and MAX485 voltage source is esp32 from 16 17 GND pin and not used termination resistor yet

Vcc to vcc is results 0 0 0 0 response

**Reply by canton7** [reply]
Make sure that your MAX-485 voltage is actually 5V, not 3V3.

**Reply by iqbaal77** [reply]
Can i give it 5v from external source?

**Reply by canton7** [reply]
What do you mean by "Vcc to vcc"?

**Reply by canton7** [reply]
Just get a multimeter out and measure the voltage you're giving to the MAX-485...

**Reply by iqbaal77** [reply]
> What do you mean by "Vcc to vcc"?

Esp32 vcc to Max485 vcc

**Reply by iqbaal77** [reply]
Now i have rs485 to USB converter and I'm trying to read 110xx 300xx 400xx registers using modscan but each time i am getting "modbus exception from slave device" i am using foxess kh8 hybrid inverter, please help me, I'm stuck!! I need its actual register list because I'm not sure which one is correct because there is alot, also i need guide after connecting with modscan what actual values i need to put for my kh8 hybrid inverter...

**Reply by FozzieUK** [reply]
That implies that you are getting a response which is invalid, which confirms what you saw in your first post

What are you using to scan the modbus is it modpoll ? - can you post what the command is you are sending.

And just to double check again you are connected to the 485A/B pins (3 & 4) and not the Meter485A/B pins.

**Reply by iqbaal77** [reply]
Yes its 100% clear I'm using rs485 3-4 not meter ones and I'm using modscan and selecting 11000 30000 40000 series of commands by selecting in modscan

**Reply by canton7** [reply]
> modbus exception from slave device

Any detail on which exception exactly you're getting?

**Reply by iqbaal77** [reply]
![image](https://github.com/user-attachments/assets/84d4c99d-1bc4-40bb-b2a7-3e57b4987eec)

![image](https://github.com/user-attachments/assets/2d56d7bd-c209-42dd-a8fb-741aeed2a0e7)

**Reply by canton7** [reply]
So exception code 02, which is "Invalid data address". An you're trying to read from address 1 there, which isn't a valid input register. Indeed, the KH (post firmware version 1.19) only uses holding registers, in the 3xxxx and 4xxxx ranges.

**Reply by canton7** [reply]
I'm going to guess the "Address" box at the top...?

Ultimately, I'm afraid that this discussion forum is about the foxess_modbus integration. We're happy to provide a limited amount of support for things unrelated to foxess_modbus, but this isn't really the right place to hand-hold you through the basics of modbus.

**Reply by canton7** [reply]
That's 3000 not 30000 (at last that's what the box at the top says, although the window at the bottom says 30000. Not sure what's going on there). And I said you need to read the holding registers, not input registers.

**Reply by iqbaal77** [reply]
iqbaal77@gmail.com

**Reply by canton7** [reply]
Are you sure you requested register 30000? Can I see your request?

**Reply by iqbaal77** [reply]
> Are you sure you requested register 30000? Can I see your request?

0xF7 0x04 0x79 0x18 0x00 0x0F 0x3D 0xC3
I requested 31000 as per manual let me share 30000 response.

**Reply by canton7** [reply]
> > Are you sure you requested register 30000? Can I see your request?
> 
> 0xF7 0x04 0x79 0x18 0x00 0x0F 0x3D 0xC3 I requested 31000 as per manual let me share 30000 response.

Why did you reply to a message saying to fetch register 30000, with a response to register 31000? How was I supposed to guess that the response you posted was for register 31000? I wasted a chunk of time trying to decode that as register 30000.

Anyway:

> 0xF7 0x04 0x1E 0x0A 0x3E 0x00 0x39 0x05 0xE1 0x0A 0x68 0x00 0x39 0x05 0xFA 0x00 0x08 0x00 0x00 0x00 0x00 0x00 0x00 0x08 0xFC 0x00 0x7C 0x00 0x00 0x13 0x87 0x00 0x00 0x24 0x27

F7: Slave ID
04: Read input registers (I did say to read holding registers, but some inverters expose registers as both)
1E: 30 bytes of response
0A3E: PV1 voltage: 2662. Scale is V\*10, so this is 266.2V  
0039: PV1 current: 57. Scale is A\*10, so this is 5.7A
05E1: PV1 power: 1505. Scale is W, so this is 1.505kW. 

And so on.

**Reply by iqbaal77** [reply]
I'm really sorry for my mistake, can you please share how I can match registers with each output?

**Reply by canton7** [reply]
I just worked through an example of decoding the first 3 registers in your response (31000, 31001, 31002). What about that example did you not understand?

**Reply by iqbaal77** [reply]
> I just worked through an example of decoding the first 3 registers in your response (31000, 31001, 31002). What about that example did you not understand?

I got that 100% just wanted to ask correct manual for all registers that you are following.

**Reply by canton7** [reply]
I'm not sure what you mean. The registers are all listed in the protocol document which you linked to at the top of this discussion.
